snapshot/snapshotfs/restore.go

```diff
@@ -4,89 +4,133 @@ import (
+		err = restoreDirectory(ctx, e, targetPath)
+	case fs.File:
+		err = restoreFile(ctx, targetPath, e)
+	case fs.Symlink:
+		// TODO:
+		return nil
+	default:
+		return errors.Errorf("Invalid FS entry type for %s: %#v", targetPath, e)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return setAttributes(targetPath, e)
+}
+
+// set permission, modification time and user/group ids on targetPath
+func setAttributes(targetPath string, e fs.Entry) error {
+	const modBits = os.ModePerm | os.ModeSetgid | os.ModeSetuid | os.ModeSticky
+
+	le, err := localfs.NewEntry(targetPath)
+	if err != nil {
+		return errors.Wrap(err, "could not create local FS entry for "+targetPath)
+	}
+
+	// Set owner user and group from e
+	if le.Owner() != e.Owner() {
+		if err = os.Chown(targetPath, int(e.Owner().UserID), int(e.Owner().GroupID)); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change owner/group for "+targetPath)
+		}
+	}
+
+	// Set file permissions from e
+	if (le.Mode() & modBits) != (e.Mode() & modBits) {
+		if err = os.Chmod(targetPath, e.Mode()&modBits); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change permissions on "+targetPath)
+		}
+	}
+
+	// Set mod time from e
+	if !le.ModTime().Equal(e.ModTime()) {
+		// Note: Set atime to ModTime as well
+		if err = os.Chtimes(targetPath, e.ModTime(), e.ModTime()); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change mod time on "+targetPath)
+		}
+	}
+
+	return nil
+}
+
+func restoreDirectory(ctx context.Context, d fs.Directory, targetPath string) error {
+	if err := createDirectory(targetPath); err != nil {
+		return err
+	}
 
 	entries, err := d.Readdir(ctx)
 	if err != nil {
 		return err
 	}
 
 	for _, e := range entries {
-		restorePath := path.Join(targetPath, e.Name())
-		objectID := e.(object.HasObjectID).ObjectID()
-		// Restore directories recursively
-		if e.Mode().IsDir() {
-			if direrr := createDirectory(restorePath); direrr != nil {
-				return direrr
-			}
-			if recerr := Restore(ctx, rep, restorePath, objectID); recerr != nil {
-				return recerr
-			}
-			// Set permissions as stored in the snapshot
-			if cherr := os.Chmod(restorePath, e.Mode()); cherr != nil && !os.IsPermission(cherr) {
-				return cherr
-			}
-			continue
-		}
-		if ferr := restoreFile(ctx, rep, restorePath, objectID, e.Mode()); ferr != nil {
-			return ferr
+		if err := restoreEntry(ctx, e, filepath.Join(targetPath, e.Name())); err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
 func createDirectory(path string) error {
-	stat, err := os.Stat(path)
-	switch {
-	case err == nil:
-		if stat.Mode().IsRegular() {
-			return errors.Errorf("unable to create directory, %s already exists as a file", path)
-		}
-		return nil
+	switch stat, err := os.Stat(path); {
 	case os.IsNotExist(err):
 		return os.MkdirAll(path, 0700)
+	case err != nil:
+		return errors.Wrap(err, "failed to stat path "+path)
+	case stat.Mode().IsDir():
```

Comment:
We may want to fail in this case as well to enforce that none of the target directories exist and avoid restoring files on an existing directory.  There is a special case for handling restores to the root directory already.

Reply:
done